File: structural-design-pattern/flyweight-pattern/flyweight.py
"""
Flyweight Pattern
=================
Category: Structural Design Pattern

Intent:
    Minimize memory usage by sharing as much data as possible between similar
    objects. The shared part is called "intrinsic state" (stored in the flyweight),
    while the unique part is "extrinsic state" (passed in by the client).

When to use:
    - When you have a huge number of similar objects consuming too much memory
    - When most object state can be made extrinsic (passed in rather than stored)
    - Common in: text editors (character rendering), game engines (particles, tiles),
      caching systems, icon/image pools

Key Participants:
    - Flyweight: Stores intrinsic (shared) state
    - FlyweightFactory: Creates and caches flyweight objects
    - Client: Maintains extrinsic state and passes it to flyweights

Intrinsic vs Extrinsic state:
    - Intrinsic: Shared, immutable data (e.g., character font, glyph bitmap)
    - Extrinsic: Unique per usage, passed in at runtime (e.g., position, color)
"""

from dataclasses import dataclass


# ---------------------------------------------------------------------------
# Flyweight — stores INTRINSIC (shared) state
# ---------------------------------------------------------------------------
@dataclass(frozen=True)  # Immutable — safe to share
class ParticleType:
    """
    Shared data for a type of game particle. A "fire" particle always uses
    the same sprite, animation, and base color regardless of where it appears.
    These are the INTRINSIC properties.
    """
    name: str
    sprite: str      # Path to sprite image (imagine this is a large texture)
    animation: str    # Animation sequence name
    base_color: str   # Default color

    def render(self, x: int, y: int, scale: float, opacity: float) -> str:
        """
        Render this particle type at a specific position.
        x, y, scale, opacity are EXTRINSIC state — unique per particle instance.
        """
        return (
            f"[{self.name}] sprite={self.sprite} at ({x},{y}) "
            f"scale={scale:.1f} opacity={opacity:.1f} color={self.base_color}"
        )


# ---------------------------------------------------------------------------
# Flyweight Factory — creates and caches flyweights
# ---------------------------------------------------------------------------
class ParticleTypeFactory:
    """
    Ensures each unique ParticleType is created only once.
    If 10,000 fire particles exist, they all share ONE ParticleType instance.
    """

    def __init__(self) -> None:
        self._cache: dict[str, ParticleType] = {}

    def get(self, name: str, sprite: str, animation: str, base_color: str) -> ParticleType:
        key = f"{name}:{sprite}:{animation}:{base_color}"
        if key not in self._cache:
            self._cache[key] = ParticleType(name, sprite, animation, base_color)
            print(f"  [Factory] Created new flyweight for '{name}'")
        return self._cache[key]

    @property
    def count(self) -> int:
        return len(self._cache)


# ---------------------------------------------------------------------------
# Client context — stores EXTRINSIC (per-instance) state
# ---------------------------------------------------------------------------
@dataclass
class Particle:
    """
    An individual particle in the game world. The heavy data (sprite, animation)
    is shared via the flyweight. Only position, scale, and opacity are unique.
    """
    particle_type: ParticleType  # Reference to shared flyweight
    x: int                       # Extrinsic: unique position
    y: int
    scale: float
    opacity: float

    def render(self) -> str:
        return self.particle_type.render(self.x, self.y, self.scale, self.opacity)


# ---------------------------------------------------------------------------
# Usage Example
# ---------------------------------------------------------------------------
if __name__ == "__main__":
    import random

    factory = ParticleTypeFactory()

    # Imagine a game creating thousands of particles
    particles: list[Particle] = []

    # Create 1000 fire particles — all share ONE ParticleType flyweight
    for _ in range(1000):
        ptype = factory.get("fire", "sprites/fire.png", "burn_loop", "#FF4500")
        particles.append(Particle(
            particle_type=ptype,
            x=random.randint(0, 1920),
            y=random.randint(0, 1080),
            scale=random.uniform(0.5, 2.0),
            opacity=random.uniform(0.3, 1.0),
        ))

    # Create 500 smoke particles — share another flyweight
    for _ in range(500):
        ptype = factory.get("smoke", "sprites/smoke.png", "fade_up", "#808080")
        particles.append(Particle(
            particle_type=ptype,
            x=random.randint(0, 1920),
            y=random.randint(0, 1080),
            scale=random.uniform(1.0, 3.0),
            opacity=random.uniform(0.1, 0.5),
        ))

    # Create 200 spark particles
    for _ in range(200):
        ptype = factory.get("spark", "sprites/spark.png", "flash", "#FFD700")
        particles.append(Particle(
            particle_type=ptype,
            x=random.randint(0, 1920),
            y=random.randint(0, 1080),
            scale=random.uniform(0.2, 0.8),
            opacity=random.uniform(0.5, 1.0),
        ))

    print(f"\nTotal particles: {len(particles)}")
    print(f"Unique flyweights (shared types): {factory.count}")
    print(f"Memory savings: instead of {len(particles)} sprite/animation copies,")
    print(f"  we only have {factory.count} (shared via flyweight references)")

    # Render a sample
    print("\nSample renders:")
    for p in particles[:5]:
        print(f"  {p.render()}")

File: structural-design-pattern/flyweight-pattern/flyweight.ts
/**
 * Flyweight Pattern
 * =================
 * Category: Structural Design Pattern
 *
 * Intent:
 *   Minimize memory usage by sharing data between similar objects.
 *   Shared data = "intrinsic state" (in the flyweight).
 *   Unique data = "extrinsic state" (passed by the client).
 *
 * When to use:
 *   - Huge number of similar objects consuming too much memory
 *   - Most object state can be externalized
 *   - Common in: text editors, game engines (particles, tiles), caching
 *
 * Key Participants:
 *   - Flyweight: Stores intrinsic (shared) state
 *   - FlyweightFactory: Creates and caches flyweight instances
 *   - Client: Maintains extrinsic state
 */

// ---------------------------------------------------------------------------
// Flyweight — shared intrinsic state (immutable)
// ---------------------------------------------------------------------------
class ParticleType {
  constructor(
    readonly name: string,
    readonly sprite: string, // Imagine this is a large texture reference
    readonly animation: string,
    readonly baseColor: string
  ) {}

  /** Extrinsic state (x, y, scale, opacity) is passed in, not stored. */
  render(x: number, y: number, scale: number, opacity: number): string {
    return `[${this.name}] sprite=${this.sprite} at (${x},${y}) scale=${scale.toFixed(1)} opacity=${opacity.toFixed(1)} color=${this.baseColor}`;
  }
}

// ---------------------------------------------------------------------------
// Flyweight Factory — ensures each type is created only once
// ---------------------------------------------------------------------------
class ParticleTypeFactory {
  private cache = new Map<string, ParticleType>();

  get(
    name: string,
    sprite: string,
    animation: string,
    baseColor: string
  ): ParticleType {
    const key = `${name}:${sprite}:${animation}:${baseColor}`;
    if (!this.cache.has(key)) {
      this.cache.set(key, new ParticleType(name, sprite, animation, baseColor));
      console.log(`  [Factory] Created new flyweight for '${name}'`);
    }
    return this.cache.get(key)!;
  }

  get count(): number {
    return this.cache.size;
  }
}

// ---------------------------------------------------------------------------
// Client context — stores extrinsic (per-instance) state
// ---------------------------------------------------------------------------
interface Particle {
  type: ParticleType; // Reference to shared flyweight
  x: number; // Extrinsic: unique position
  y: number;
  scale: number;
  opacity: number;
}

function renderParticle(p: Particle): string {
  return p.type.render(p.x, p.y, p.scale, p.opacity);
}

// ---------------------------------------------------------------------------
// Usage Example
// ---------------------------------------------------------------------------
const factory = new ParticleTypeFactory();
const particles: Particle[] = [];

// 1000 fire particles — all share ONE ParticleType flyweight
for (let i = 0; i < 1000; i++) {
  particles.push({
    type: factory.get("fire", "sprites/fire.png", "burn_loop", "#FF4500"),
    x: Math.floor(Math.random() * 1920),
    y: Math.floor(Math.random() * 1080),
    scale: 0.5 + Math.random() * 1.5,
    opacity: 0.3 + Math.random() * 0.7,
  });
}

// 500 smoke particles
for (let i = 0; i < 500; i++) {
  particles.push({
    type: factory.get("smoke", "sprites/smoke.png", "fade_up", "#808080"),
    x: Math.floor(Math.random() * 1920),
    y: Math.floor(Math.random() * 1080),
    scale: 1.0 + Math.random() * 2.0,
    opacity: 0.1 + Math.random() * 0.4,
  });
}

// 200 spark particles
for (let i = 0; i < 200; i++) {
  particles.push({
    type: factory.get("spark", "sprites/spark.png", "flash", "#FFD700"),
    x: Math.floor(Math.random() * 1920),
    y: Math.floor(Math.random() * 1080),
    scale: 0.2 + Math.random() * 0.6,
    opacity: 0.5 + Math.random() * 0.5,
  });
}

console.log(`\nTotal particles: ${particles.length}`);
console.log(`Unique flyweights (shared types): ${factory.count}`);
console.log(
  `Memory savings: ${particles.length} particles share only ${factory.count} type objects`
);

console.log("\nSample renders:");
particles.slice(0, 5).forEach((p) => console.log(`  ${renderParticle(p)}`));
